### PR TITLE
25859 lib lite

### PR DESCRIPTION
**Pull request by thekarinka**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,153 @@
...
+
+The javascript library **MojeID LITE** (or mojeID Connect) allows loading myID data into a client-side web page using the OpenID Connect protocol.
+
+This functionality can be used, for example, to easily pre-fill the web form with the details of a user who has an active mojeID account.
+
+Description of the library (in Czech) can be found on the page [4.1.11. Knihovna MojeID LITE](https://www.mojeid.cz/dokumentace/html/ImplementacePodporyMojeid/OpenidConnect/KnihovnaJS.html).
+
+### Installation
+
+Create folder for `npm` packages.
```
Comment (thekarinka) [inline README.md]: This is not required, folder can be created also automatically.

File: AUTHORS [hunk 2]
```diff
@@ -0,0 +1,3 @@
+Zdeněk Böhm <zdenek.bohm@nic.cz>
+Tomáš Pazderka <tomas.pazderka@nic.cz>
+Michal Vodička
```
Comment (thekarinka) [inline AUTHORS]: missing new line

File: .eslintrc.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        "keyword-spacing":      ["warn"],                   // Reqiure spacing around keywords
+        "func-call-spacing":    ["error", "never"],         // Don't allow spaces between function call and arguments
+        "space-before-function-paren": ["error", "never"],  // Don't allow spaces before function definition parenthesis
+        "spaced-comment":       ["warn", "always"],         // Require whitespace after //
+        "space-in-parens":      ["warn", "never"],          // Don't allow spaces in parenthesis
+        "comma-spacing":        ["warn"],                   // Reqire space after comma, don't allow space before
+        "space-infix-ops":      ["warn"],                   // Require spaces around infix operators (including =)
+        "no-console":           ["off"]
+    }
+};
```
Comment (thekarinka) [inline .eslintrc.js]: missing new line

File: examples/dist [hunk 4]
```diff
@@ -0,0 +1 @@
+../dist/
```
Comment (thekarinka) [inline examples/dist]: Why is there need this link?

File: examples/node_modules [hunk 5]
```diff
@@ -0,0 +1 @@
+../node_modules
```
Comment (thekarinka) [inline examples/node_modules]: I think that better solution is create two separete project, than llinking node modules folder.

File: package.json [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    "test": "ava",
+    "test-coverage": "nyc ava",
+    "build": "webpack"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/zbohm/mojeid-connect-js.git"
+  },
+  "author": "Zdeněk Böhm <zdenek.bohm@nic.cz>",
+  "license": "ISC",
```
Comment (thekarinka) [inline package.json]: There is mentioned another license beside license file in repository.

File: package.json [hunk 7]
```diff
@@ -0,0 +1,74 @@
+{
+  "name": "@cz-nic/mojeid-connect",
+  "version": "2.0.0",
+  "description": "MojeID LITE Connection library",
+  "main": "dist/mojeid_connect.2.0.0.js",
+  "directories": {
+    "example": "examples"
+  },
+  "scripts": {
```
Comment (thekarinka) [inline package.json:9]: It could be cool add also watch mode.

File: examples/form-callback.html [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+  <meta charset="UTF-8">
+  <title>Form callback example</title>
+  <script src="/dist/mojeid_connect.2.0.0.js" id="mojeid-connect-script" data-jsrsasign="/node_modules/jsrsasign/lib/jsrsasign-all-min.js"></script>
+  <script>
+    document.addEventListener("DOMContentLoaded", function () {
+
+      var mojeID = mojeid_connect.createMojeidConnect({
+        clientName: "Form callback example",
+        clientIdName: "form_callback_example",
+        formCallback: mojeid_connect.getFillFormCallback("input-name"),
```
Comment (thekarinka) [inline examples/form-callback.html]: This line required more effort to understand what it do. I would prefer modify getFillFormCallback:
- use in paramaeter default value for name as `name="input_name"`
- in cases where is need match elemensts by attribute name you will call it without arguments (and you do not need to thinking about "what the paramaters does do"), in other cases you will use custom value

File: src/client.js [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+    }
+
+    /**
+     * Set value to elements.
+     *
+     * @param {List} nodes
+     * @param {String} value
+     */
+    setValueToElements(nodes, value) {
+        for (let i = 0; i < nodes.length; i++) {
```
Comment (thekarinka) [inline src/client.js]: This could be simplified by using for(const input of nodes)

File: src/index.js [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * @function createMojeidConnect
+ * @returns {function} fillFormCallback
+ */
+export function getFillFormCallback(name) {
+    if (name === "input-name") {
+        return fillFormQuerySelectorAllInputName
+    }
+    return fillFormElementById
+}
```
Comment (thekarinka) [inline src/index.js]: missing newline

File: test/test_client.js [hunk 11]
```diff
@@ -0,0 +1,406 @@
+import test from 'ava'
```
Comment (thekarinka) [inline test/test_client.js:1]: Could you add config for travis with check code style and tests running?

File: examples/form-callback.html [hunk 12]
```diff
@@ -0,0 +1,80 @@
+<!DOCTYPE html>
+<html lang="en">
+ <head>
+  <meta charset="UTF-8">
+  <title>Form callback example</title>
+  <script src="/dist/mojeid_connect.2.0.0.js" id="mojeid-connect-script" data-jsrsasign="/node_modules/jsrsasign/lib/jsrsasign-all-min.js"></script>
```
Comment (thekarinka) [inline examples/form-callback.html:6]: In general, maybe we could think about using https://requirejs.org/
It minimalize javascript in html files and also I think that we could use it for jsrslib loading.
But this is to think about impacts, possitives and negatives of using this solution.

**Comment by thekarinka** [comment]
Also, there is missing ChangeLog.rst

File: README.md (hunk 1 above)
Comment (zbohm) [reply README.md]: The paragraph removed.

File: AUTHORS (hunk 2 above)
Comment (zbohm) [reply AUTHORS]: Line added.

File: .eslintrc.js (hunk 3 above)
Comment (zbohm) [reply .eslintrc.js]: Line added.

File: examples/dist (hunk 4 above)
Comment (zbohm) [reply examples/dist]: The link removed. It was designated only for testing examples directly from the git project.

File: examples/node_modules (hunk 5 above)
Comment (zbohm) [reply examples/node_modules]: The link removed like the `dist`. For now we will keep examples in this project.

File: examples/form-callback.html (hunk 8 above)
Comment (zbohm) [reply examples/form-callback.html]: This function has been removed due to redundancy.

File: src/client.js (hunk 9 above)
Comment (zbohm) [reply src/client.js]: The code simplified.

File: src/index.js (hunk 10 above)
Comment (zbohm) [reply src/index.js]: Line added.

File: test/test_client.js (hunk 11 above)
Comment (zbohm) [reply test/test_client.js:1]: Config added.

File: examples/form-callback.html (hunk 12 above)
Comment (zbohm) [reply examples/form-callback.html:6]: I wouldn't use it yet. I'd like to keep the project as simple as possible.

**Review by zbohm — COMMENTED** [review]
Changelog added.

File: .eslintrc.js [hunk 13]
```diff
@@ -28,7 +28,7 @@ module.exports = {
         "prefer-const":         ["warn", { "ignoreReadBeforeAssign": true }],  // If let is not assigned to, prefer const
         "no-return-assign":     ["error", "always"],        // Don't allow assingment in return statement
         "eqeqeq":               ["error", "smart"],         // Reqire use of === and !== instead of == and != (with a few exceptions)
-        "complexity":           ["warn", 20],               // Maximum cyclomatic complexity
+        "complexity":           ["warn", 28],               // Maximum cyclomatic complexity
```
Comment (thekarinka) [inline .eslintrc.js]: Is there some specific reason why we are increasing cyclomatic complexity? As I know, our longterm effort is decresing complexity under 15 by refactoring and creating new code.

File: CHANGELOG.md [hunk 14]
```diff
@@ -0,0 +1,10 @@
+# Changelog
```
Comment (thekarinka) [inline CHANGELOG.md]: ChangeLog should be wrriten in rst format, see https://trac.nic.cz/wiki/sqa/documenting/repositories/changelog for more details

File: examples/all-attrs.html [hunk 15]
```diff
@@ -0,0 +1,369 @@
+<!DOCTYPE html>
+<html lang="en">
+ <head>
+  <meta charset="UTF-8">
+  <title>Request for all attributes</title>
+  <script src="/dist/mojeid_connect.2.0.0.js" id="mojeid-connect-script" data-jsrsasign="/node_modules/jsrsasign/lib/jsrsasign-all-min.js"></script>
+  <script>
+    document.addEventListener("DOMContentLoaded", function () {
```
Comment (thekarinka) [inline examples/all-attrs.html:8]: I propose to exclude this code to separate file, rewrite it in ES6 and also transpile it.

File: examples/all-attrs.html [hunk 16]
```diff
@@ -0,0 +1,369 @@
...
+      <div><label>birthdate:</label><input type="text" id="birthdate"/></div>
+      <div><label>gender:</label><input type="text" id="gender"/></div>
+      <div><label>phone_number:</label><input type="text" id="phone_number"/></div>
+      <div><label>profile:</label><input type="text" id="profile"/></div>
+      <div><label>website:</label><input type="text" id="website"/></div>
+      <div><label>address:</label><textarea cols="32" rows="5" id="address"></textarea></div>
+    </section>
+
+    <section>
+        <h3>Identification data</h3>
```
Comment (thekarinka) [inline examples/all-attrs.html:260]: I propose to format this code, some lines are hard to read. At least something like this:
```
<div>
    <label ...>
    <input ...>
</div>
```

File: examples/all-attrs.html [hunk 17]
```diff
@@ -0,0 +1,369 @@
...
+        <div><label>mojeid_address_def_street3:</label><input type="text" id="mojeid_address_def_street3"/></div>
+        <div><label>mojeid_address_def_city:</label><input type="text" id="mojeid_address_def_city"/></div>
+        <div><label>mojeid_address_def_state:</label><input type="text" id="mojeid_address_def_state"/></div>
+        <div><label>mojeid_address_def_postal_code:</label><input type="text" id="mojeid_address_def_postal_code"/></div>
+        <div><label>mojeid_address_def_country:</label><input type="text" id="mojeid_address_def_country"/></div>
+    </section>
+
+    <section>
+        <h3>Mail address</h3>
+        <!--
```
Comment (thekarinka) [inline examples/all-attrs.html]: Why is there this comment?

File: examples/form-callback.html [hunk 18]
```diff
@@ -7,10 +7,16 @@
   <script>
     document.addEventListener("DOMContentLoaded", function () {
 
+      function setMessage(class_name="", text="") {
```
Comment (thekarinka) [inline examples/form-callback.html:10]: Same as above. I think it is better to exclude to separate file.

File: examples/parse-claims.html [hunk 19]
```diff
@@ -6,6 +6,12 @@
   <script>
     document.addEventListener("DOMContentLoaded", function () {
 
+      function setMessage(class_name="", text="") {
```
Comment (thekarinka) [inline examples/parse-claims.html:10]: Against, separate file, also I think that there is duplicate code.

File: src/index.js [hunk 20]
```diff
@@ -1,31 +1,15 @@
-import MojeidConnect from './client'
-import {fillFormElementById, fillFormQuerySelectorAllInputName} from './utils'
-
+import MojeidConnect, {defaultClientIdName} from './client'
+export {fillFormQuerySelectorAllInputName} from './utils'
```
Comment (thekarinka) [inline src/index.js]: export?

File: README.md [hunk 21]
```diff
@@ -145,8 +142,14 @@ Compile the library. Files are compiled into the `./dist` folder.
 npm run build
 ```
 
-Follow the instructions from the point [Create self signed SSL certificate](#create-self-signed-ssl-certificate).
-(Only `path` in command `http-server` will be `./examples/`.)
+Create certificate as it is described in [Create self signed SSL certificate](#create-self-signed-ssl-certificate). Create symlinks to the distributed library and node modules (only once) and run server:
+
+```
+ln -s ../node_modules examples/node_modules
```
Comment (thekarinka) [inline README.md:148]: Is there really need these links? From the doc https://nodejs.org/api/modules.html#modules_loading_from_node_modules_folders looks like nodejs is able to find node_modules also in parent folder, it doesn't?
Also, dist folder could be removed by some additions in webpack config, couldn't?

File: CHANGELOG.md (hunk 14 above)
Comment (zbohm) [reply CHANGELOG.md]: This is changelog for Trac. I use the version https://keepachangelog.com/en/1.0.0/ as is user for example at https://github.com/olivierlacan/keep-a-changelog/blob/master/CHANGELOG.md

File: src/index.js (hunk 20 above)
Comment (zbohm) [reply src/index.js]: Yes. It is not public function.

File: CHANGELOG.md (hunk 14 above)
Comment (stinovlas) [reply CHANGELOG.md]: It's not changelog for Trac. It's our company convention for writing changelogs in all our projects.

File: package.json (hunk 6 above)
Comment (zbohm) [reply package.json]: Fixed.

File: package.json (hunk 7 above)
Comment (zbohm) [reply package.json:9]: watch added

File: .eslintrc.js (hunk 13 above)
Comment (zbohm) [reply .eslintrc.js]: Change canceled.

File: examples/all-attrs.html (hunk 16 above)
Comment (zbohm) [reply examples/all-attrs.html:260]: Examples formatted.

File: examples/all-attrs.html (hunk 17 above)
Comment (zbohm) [reply examples/all-attrs.html]: Comment removed.

File: CHANGELOG.md (hunk 14 above)
Comment (zbohm) [reply CHANGELOG.md]: Renamed and revritten according of CZ.NIC conventions.

File: examples/all-attrs.html (hunk 15 above)
Comment (zbohm) [reply examples/all-attrs.html:8]: I don't thik this is a good idea. The example should be simple and in one place. So no one has to look for it and fold it up.

File: README.md (hunk 21 above)
Comment (zbohm) [reply README.md:148]: The library is written to by included into html code. There is also a function to load jsrsasign from html. That wouldn't make sense by using `requrire`.

**Comment by zbohm** [comment]
All ambiguities explained and resolved. The branch was rebased and is ready to merge.

**Pull request closed without merging**